File: day07/solver.py
#!/usr/bin/env python

from pathlib import Path
from aoc import utils
from operator import add, mul

watch = utils.stopwatch()


@watch.measure_time
def parse(raw_data):
    equations = []
    for line in raw_data.splitlines():
        nums = utils.ints(line)
        equations.append((nums[0], nums[1:]))
    return equations


def check(test, nums, operators):
    def _check(test, nums, acc):
        if not nums:
            return test == acc
        if acc > test:
            return False
        return any(_check(test, nums[1:], op(acc, nums[0])) for op in operators)

    return _check(test, nums[1:], nums[0])


@watch.measure_time
def solve1(data):
    s = 0
    for test, nums in data:
        if check(test, nums, [add, mul]):
            s += test
    return s


def combine_digits(a, b):
    return int(str(a) + str(b))


@watch.measure_time
def solve2(data):
    s = 0
    for test, nums in data:
        if check(test, nums, [add, mul, combine_digits]):
            s += test
    return s


if __name__ == "__main__":
    data = parse(open(Path(__file__).parent / "input.txt").read().strip())
    print(f"Part 1: {solve1(data)}")
    print(f"Part 2: {solve2(data)}")
    print()
    watch.print_times()

File: day07/solver.hs
{-# LANGUAGE QuasiQuotes #-}

import Text.RawString.QQ 
import Data.List
import Utils
import AoC

getInput :: IO String
getInput = readFile "input.txt"

--------------------------------------------------------------------

type Parsed = [(Int, [Int])]
type Sol1 = Int
type Sol2 = Int

parse :: String -> Parsed
parse = map (\ns -> (head ns, tail ns)) . map parseInts . lines


check :: [Int -> Int -> Int] -> Int -> [Int] -> Bool
check ops test (n:ns) = check' test ns n
    where
        check' test [] acc = test == acc
        check' test (n:ns) acc
            | acc > test = False
            | otherwise = any (\op -> check' test ns (op acc n)) ops


solve1 :: Parsed -> Sol1
solve1 = sum . map fst . filter (uncurry (check [(+), (*)]))


concatInts :: Int -> Int -> Int
concatInts a b = read $ show a ++ show b


solve2 :: Parsed -> Sol2
solve2 = sum . map fst . filter (uncurry (check [(+), (*), concatInts]))


testdata = [r||]
testresult1 = 0
testresult2 = 0

--------------------------------------------------------------------

test1 = test (solve1 . parse) testresult1 testdata
test2 = test (solve2 . parse) testresult2 testdata

printPart :: (Show a) => Int -> a -> IO ()
printPart part solution = do
    putStr $ "Part " ++ (show part) ++ ": "
    print solution

main' :: CLIOptions -> IO ()
main' (CLIOptions p False) = do
    parsed <- parse <$> getInput
    case p of
        1 -> printPart 1 (solve1 parsed)
        2 -> printPart 2 (solve2 parsed)
        _ -> (do
            printPart 1 (solve1 parsed)
            printPart 2 (solve2 parsed))
main' (CLIOptions p True) = 
    case p of
        1 -> printPart 1 test1
        2 -> printPart 2 test2
        _ -> (do
            printPart 1 test1
            printPart 2 test2)

main = main' =<< clioptions
